### PR TITLE
Make scraping offset consistent.

### PR DESCRIPTION
**Pull request by fabxc**
To evenly distribute scraping load we currently rely on random
jittering. This commit hashes over the target's identity and calculates
a consistent offset. This also ensures that scrape intervals
are constantly spaced between config/target changes.

### CODE REVIEW

File: retrieval/target.go [hunk 1]
```diff
@@ -267,6 +266,36 @@ func (t *Target) String() string {
...
 }
 
+// fingerprint returns an identifying hash for the target.
+func (t *Target) fingerprint() model.Fingerprint {
+	t.RLock()
+	defer t.RUnlock()
+
+	lset := t.Labels()
+
+	lset[model.MetricsPathLabel] = t.labels[model.MetricsPathLabel]
```
Comment (brian-brazil) [inline retrieval/target.go]: I think we should include all labels. This right now would clash on snmp/blackbox as it's one target with varying params.

File: retrieval/target.go [hunk 2]
```diff
@@ -267,6 +266,36 @@ func (t *Target) String() string {
...
+	var (
+		base   = now - (now % int64(interval))
+		offset = uint64(t.fingerprint()) % uint64(interval)
+		next   = base + int64(offset)
+	)
+
+	if next < now {
+		next += int64(interval)
+	}
+	return time.Duration(next - now)
```
Comment (beorn7) [inline retrieval/target.go]: You add `now` in L288. Then you substract it here.

Just leave it out in both places and compare `next < 0` in L293. Should have the same effect.

**Comment by beorn7** [comment]
Cool idea.

**Comment by fabxc** [comment]
Both comments fixed.

File: retrieval/target.go [hunk 3]
```diff
@@ -267,6 +266,30 @@ func (t *Target) String() string {
...
+
+	return t.labels.Fingerprint()
+}
+
+// offset returns the time until the next scrape cycle for the target.
+func (t *Target) offset(interval time.Duration) time.Duration {
+	now := time.Now().UnixNano()
+
+	var (
+		base   = now % int64(interval)
```
Comment (brian-brazil) [inline retrieval/target.go]: Shouldn't this be a division rather than a modulus?

Comment (fabxc) [inline retrieval/target.go]: That's essentially the base offset. To that we add another offeset between 0 and `interval` based on the fingerprint.
Finally, if this total offest (`next`) falls behind now, we move it one cycle ahead.

The returned duration is the wait time, not the absolute time of the next scrape.

Comment (beorn7) [inline retrieval/target.go]: To be equivalent to before, you needed `-(now % int64(interval)` here.

Or you check in L287 if `next > interval` (and then substract interval). The latter might be most intuitive.

Comment (fabxc) [inline retrieval/target.go]: Fixed.

**Comment by brian-brazil** [comment]
:+1:

**Comment by beorn7** [comment]
:+1:

**Pull request merged**